src/array.cr

```diff
@@ -301,6 +301,16 @@ class Array(T)
   # [1, 2, 3] - [2, 1] # => [3]
   # ```
   def -(other : Array(U)) forall U
+    # Heurisitic: if any of the arrays is small we just do a linear scan
+    # instead of using a Hash for lookup.
+    if size <= SMALL_ARRAY_SIZE || other.size <= SMALL_ARRAY_SIZE
```

Comment:
Ditto: `size + other.size <= SMALL_ARRAY_SIZE * 2`

Reply:
This one, I'm not sure. I think the idea of `||` here is that if you have an array with 2 elements and another with 2_000, it'll still be faster to do 2 * 2_000 checks instead of either 1) building a hash of 2 elements and checking that against 2_000 elements, or building a hash of 2_000 elements (potentially less) and checking each against 2 elements. If either of the arrays is small it's better to do a linear scan instead of a scan in a Hash.

Reply:
Yeah, you're right!

Reply:
Is the linear scan on a small array faster than the hash lookup or are we just trying to offset the cost of building the lookup hash here? If the scan is faster than the lookup the size of `self` shouldn't matter, if `other` is small we should do the scan path. If the lookup is faster, and it's just expensive to build the hash vs a couple of scans of `other`, the size of `other` shouldn't matter, if `self` is small we should do the scan.

Reply:
@jhass Good observation!

I fiddled with the sizes to try to come up with a conclusion... Here's my code:

```crystal
require "benchmark"
require "./old_array"

sizes = [5, 10, 16, 50, 100, 500, 1_000]

sizes.each do |first_size|
  sizes.each do |second_size|
    old_a1 = OldArray.new(first_size) { Random::DEFAULT.hex(32) }
    old_a2 = OldArray.new(second_size) { |i| old_a1[i]? || Random::DEFAULT.hex(32) }

    new_a1 = Array.new(first_size) { |i| old_a1[i] }
    new_a2 = Array.new(second_size) { |i| old_a2[i] }

    puts "#{first_size} vs. #{second_size}"
    Benchmark.ips do |x|
      x.report("old") { old_a1 - old_a2 }
      x.report("new") { new_a1 - new_a2 }
    end
  end
end
```

The results are (long):

```
5 vs. 5
old   2.95M (339.39ns) (± 4.44%)   416B/op   6.86× slower
new  20.20M ( 49.50ns) (± 2.19%)  32.0B/op        fastest
5 vs. 10
old   1.51M (661.60ns) (± 2.06%)   864B/op  13.68× slower
new  20.68M ( 48.36ns) (± 0.93%)  32.0B/op        fastest
5 vs. 16
old 989.61k (  1.01µs) (± 1.46%)   864B/op  20.35× slower
new  20.14M ( 49.65ns) (± 0.75%)  32.0B/op        fastest
5 vs. 50
old 320.80k (  3.12µs) (± 0.89%)  3.85kB/op  64.39× slower
new  20.66M ( 48.41ns) (± 0.91%)   32.0B/op        fastest
5 vs. 100
old 163.54k (  6.11µs) (± 0.65%)  7.13kB/op  129.31× slower
new  21.15M ( 47.29ns) (± 0.82%)   32.0B/op         fastest
5 vs. 500
old  34.26k ( 29.19µs) (± 1.93%)  28.5kB/op  570.83× slower
new  19.56M ( 51.14ns) (± 1.29%)   32.0B/op         fastest
5 vs. 1000
old  17.21k ( 58.11µs) (± 0.83%)  56.5kB/op  1154.75× slower
new  19.87M ( 50.32ns) (± 1.26%)   32.0B/op          fastest

10 vs. 5
old   2.12M (472.47ns) (± 0.74%)  512B/op   2.62× slower
new   5.55M (180.30ns) (± 3.62%)  128B/op        fastest
10 vs. 10
old   1.26M (792.95ns) (± 0.99%)   864B/op   5.95× slower
new   7.50M (133.34ns) (± 1.70%)  32.0B/op        fastest
10 vs. 16
old 932.92k (  1.07µs) (± 1.11%)   864B/op   7.47× slower
new   6.97M (143.44ns) (± 5.82%)  32.0B/op        fastest
10 vs. 50
old 286.69k (  3.49µs) (± 4.37%)  3.85kB/op  25.21× slower
new   7.23M (138.38ns) (± 2.38%)   32.0B/op        fastest
10 vs. 100
old 150.43k (  6.65µs) (± 3.42%)  7.13kB/op  48.48× slower
new   7.29M (137.13ns) (± 2.79%)   32.0B/op        fastest
10 vs. 500
old  33.34k ( 29.99µs) (± 1.06%)  28.5kB/op  222.71× slower
new   7.43M (134.67ns) (± 2.94%)   32.0B/op         fastest
10 vs. 1000
old  17.10k ( 58.47µs) (± 2.99%)  56.5kB/op  429.73× slower
new   7.35M (136.05ns) (± 3.08%)   32.0B/op         fastest

16 vs. 5
old   1.55M (644.06ns) (± 3.02%)  624B/op   2.11× slower
new   3.28M (305.29ns) (± 2.37%)  240B/op        fastest
16 vs. 10
old 997.48k (  1.00µs) (± 2.56%)  960B/op   2.88× slower
new   2.87M (348.06ns) (± 2.97%)  128B/op        fastest
16 vs. 16
old 813.76k (  1.23µs) (± 2.88%)   864B/op   3.79× slower
new   3.08M (324.54ns) (± 3.63%)  32.0B/op        fastest
16 vs. 50
old 280.12k (  3.57µs) (± 3.37%)  3.85kB/op  11.08× slower
new   3.10M (322.09ns) (± 2.85%)   32.0B/op        fastest
16 vs. 100
old 145.79k (  6.86µs) (± 3.66%)  7.13kB/op  21.23× slower
new   3.10M (323.06ns) (± 3.54%)   32.0B/op        fastest
16 vs. 500
old  33.25k ( 30.08µs) (± 3.61%)  28.5kB/op  95.23× slower
new   3.17M (315.87ns) (± 2.16%)   32.0B/op        fastest
16 vs. 1000
old  15.95k ( 62.70µs) (± 5.31%)  56.5kB/op  176.38× slower
new   2.81M (355.45ns) (± 6.19%)   32.0B/op         fastest

50 vs. 5
old 694.63k (  1.44µs) (± 2.18%)  1.25kB/op   1.33× slower
new 922.55k (  1.08µs) (± 6.93%)    896B/op        fastest
50 vs. 10
old 454.78k (  2.20µs) (± 5.20%)  1.69kB/op   1.39× slower
new 634.36k (  1.58µs) (± 2.69%)    896B/op        fastest
50 vs. 16
old 382.91k (  2.61µs) (± 6.22%)  1.69kB/op   1.23× slower
new 470.42k (  2.13µs) (± 4.27%)    896B/op        fastest
50 vs. 50
old 223.02k (  4.48µs) (± 4.92%)  3.85kB/op   1.33× slower
new 296.48k (  3.37µs) (± 2.88%)   32.0B/op        fastest
50 vs. 100
old 143.72k (  6.96µs) (± 0.69%)  7.13kB/op   2.07× slower
new 296.78k (  3.37µs) (± 3.27%)   32.0B/op        fastest
50 vs. 500
old  31.99k ( 31.26µs) (± 4.35%)  28.5kB/op   9.07× slower
new 290.12k (  3.45µs) (± 5.24%)   32.0B/op        fastest
50 vs. 1000
old  16.15k ( 61.90µs) (± 5.49%)  56.5kB/op  18.05× slower
new 291.61k (  3.43µs) (± 4.38%)   32.0B/op        fastest

100 vs. 5
old 389.51k (  2.57µs) (± 2.04%)  2.03kB/op   1.23× slower
new 479.92k (  2.08µs) (± 5.15%)  1.66kB/op        fastest
100 vs. 10
old 295.81k (  3.38µs) (± 2.65%)  2.47kB/op   1.04× slower
new 307.86k (  3.25µs) (± 3.14%)  1.66kB/op        fastest
100 vs. 16
old 244.93k (  4.08µs) (± 5.08%)  2.47kB/op        fastest
new 217.26k (  4.60µs) (± 5.42%)  1.66kB/op   1.13× slower
100 vs. 50
old 170.70k (  5.86µs) (± 3.03%)  5.47kB/op        fastest
new  95.64k ( 10.46µs) (± 4.32%)  1.66kB/op   1.78× slower
100 vs. 100
old 121.37k (  8.24µs) (± 2.37%)  7.13kB/op        fastest
new  75.38k ( 13.27µs) (± 2.28%)   32.0B/op   1.61× slower
100 vs. 500
old  32.05k ( 31.20µs) (± 1.46%)  28.5kB/op   2.39× slower
new  76.65k ( 13.05µs) (± 2.25%)   32.0B/op        fastest
100 vs. 1000
old  16.61k ( 60.22µs) (± 2.02%)  56.5kB/op   4.61× slower
new  76.50k ( 13.07µs) (± 2.48%)   32.0B/op        fastest

500 vs. 5
old  89.65k ( 11.16µs) (± 3.90%)  13.1kB/op   1.16× slower
new 103.95k (  9.62µs) (± 1.41%)  12.7kB/op        fastest
500 vs. 10
old  70.80k ( 14.12µs) (± 0.53%)  13.5kB/op        fastest
new  67.63k ( 14.79µs) (± 0.96%)  12.7kB/op   1.05× slower
500 vs. 16
old  63.44k ( 15.76µs) (± 0.64%)  13.5kB/op        fastest
new  42.20k ( 23.70µs) (± 1.63%)  12.7kB/op   1.50× slower
500 vs. 50
old  57.43k ( 17.41µs) (± 0.69%)  16.5kB/op        fastest
new  15.96k ( 62.64µs) (± 1.74%)  12.7kB/op   3.60× slower
500 vs. 100
old  49.47k ( 20.21µs) (± 0.74%)  19.8kB/op        fastest
new   8.63k (115.88µs) (± 1.52%)  12.7kB/op   5.73× slower
500 vs. 500
old  24.55k ( 40.74µs) (± 2.01%)  28.5kB/op        fastest
new   3.08k (324.51µs) (± 2.36%)   32.0B/op   7.97× slower
500 vs. 1000
old  15.02k ( 66.59µs) (± 1.76%)  56.5kB/op        fastest
new   3.07k (325.43µs) (± 2.64%)   32.0B/op   4.89× slower

1000 vs. 5
old  44.64k ( 22.40µs) (± 0.58%)  25.1kB/op   1.07× slower
new  47.60k ( 21.01µs) (± 0.72%)  24.7kB/op        fastest
1000 vs. 10
old  37.32k ( 26.79µs) (± 0.70%)  25.5kB/op        fastest
new  30.81k ( 32.46µs) (± 1.51%)  24.7kB/op   1.21× slower
1000 vs. 16
old  33.64k ( 29.73µs) (± 1.18%)  25.5kB/op        fastest
new  21.45k ( 46.62µs) (± 1.36%)  24.7kB/op   1.57× slower
1000 vs. 50
old  31.70k ( 31.54µs) (± 0.65%)  28.5kB/op        fastest
new   7.77k (128.64µs) (± 1.56%)  24.7kB/op   4.08× slower
1000 vs. 100
old  28.59k ( 34.98µs) (± 1.10%)  31.8kB/op        fastest
new   4.04k (247.65µs) (± 1.38%)  24.7kB/op   7.08× slower
1000 vs. 500
old  17.69k ( 56.54µs) (± 1.51%)  41.1kB/op        fastest
new 965.01  (  1.04ms) (± 2.97%)  12.7kB/op  18.33× slower
1000 vs. 1000
old  12.50k ( 80.00µs) (± 1.09%)  56.5kB/op        fastest
new 752.74  (  1.33ms) (± 1.66%)   32.0B/op  16.61× slower
```

Seeing each side (left and right) separately we have:

```
Left size:
 - 5 always faster
 - 10 always faster
 - 16 always faster
 - 50 always faster
 - 100 mixed
 - 500 slower
 - 1000 slower

Right size:
 - 5 always faster
 - 10 slower when left size is >= 500
 - 16 slower when left size is >= 100
 - 50 slower when left size is >= 100
 - 100 slower when left size is >= 100
 - 500 slower when left size is >= 500
 - 1000 slower when left size is >= 500
```

Now, this is with strings, I don't know how it behaves for other element types (I think more complex objects will require more time).

But it seems if left size <= 16 then it's always a win. For the right side it's not clear, but maybe if we go with size <= 8 then it's always a win. I know for left size <= 50 it still seems a win, but I can't know if that's true for types other than String.

On the other hand, just copying Ruby's logic (the current one) might be good too. In the cases where the right side is 16 and the left side is big it's slower but not by much, and memory consumption is a bit better. So maybe it might still be worth it to simplify the code (I know it's just a division but why the asymmetry? And maybe it does work fine for ints or other types?)

Reply:
Thanks a lot for the investigation. I agree, let's keep it simple for now given we're in pretty small runtimes anyway.